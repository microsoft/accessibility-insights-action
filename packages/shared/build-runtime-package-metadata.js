// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
const { execSync } = require('child_process');
const fs = require('fs');
const path = require('path');
const sharedPackageJson = require('./package.json');
const rootPackageJson = require('../../package.json');
const rootDirectory = path.resolve(path.join(__dirname, '..', '..'));

// This is for forward compat; the latter was deprecated in favor of the former in Node 14
const rmdirSync = fs.rmSync || fs.rmdirSync;

// Produces a package.json and yarn.lock in outputDirectory suitable for runtime installation
// of the packages listed in the "externals" property of webpackConfig. The produced package
// metadata files will use the versions from the provided package's packageJson content, falling
// back to versions from this shared package, and will respect yarn resolutions present in the
// root package.json.
function buildRuntimePackageMetadata({ packageJson, webpackConfig, outputDirectory }) {
    console.log('building runtime package metadata in ${outputDirectory}');

    const packageDependencies = packageJson.dependencies;
    const sharedDependencies = sharedPackageJson.dependencies;

    const webpackExternals = webpackConfig.externals ? webpackConfig.externals : [];
    const runtimeDependencies = {};

    webpackExternals.forEach((packageName) => {
        if (packageDependencies.hasOwnProperty(packageName)) {
            runtimeDependencies[packageName] = packageDependencies[packageName];
        } else if (sharedDependencies.hasOwnProperty(packageName)) {
            runtimeDependencies[packageName] = sharedDependencies[packageName];
        } else {
            throw new Error(`Could not find any package.json entry for webpack external '${packageName}'`);
        }
    });

    const outPackageJsonPath = path.join(outputDirectory, 'package.json');
    const outPackageJsonContent = {
        ...packageJson,
        scripts: undefined,
        dependencies: runtimeDependencies,
        devDependencies: undefined,
        resolutions: rootPackageJson.resolutions,
    };

    fs.writeFileSync(outPackageJsonPath, JSON.stringify(outPackageJsonContent, undefined, 4));
    console.log(`built ${outPackageJsonPath}`);

    const rootYarnLockPath = path.join(rootDirectory, 'yarn.lock');
    const outYarnLockPath = path.join(outputDirectory, 'yarn.lock');
    fs.copyFileSync(rootYarnLockPath, outYarnLockPath);
    console.log(`seeded ${outYarnLockPath} from ${rootYarnLockPath}`);

    console.log('updating yarn.lock based on built package.json');
    // This intentionally uses execSync rather than the normally-preferred execFileSync
    // because it relies on shell behavior to pick whether to invoke yarn, yarn.exe,
    // yarn.bat, or yarn.cmd (any of these are possible depending on OS + how Yarn is
    // installed). This doesn't create a shell injection concern because the command
    // is a fixed string.
    execSync('yarn install --prod --ignore-engines --ignore-scripts', {
        stdio: 'inherit',
        cwd: outputDirectory,
    });

    const outNodeModulesPath = path.join(outputDirectory, 'node_modules');
    console.log(`removing ${outNodeModulesPath} left behind by yarn.lock update`);
    rmdirSync(outNodeModulesPath, { recursive: true });

    console.log('building runtime package metadata complete');
}

module.exports = { buildRuntimePackageMetadata };
