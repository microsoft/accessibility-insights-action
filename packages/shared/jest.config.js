// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
const baseConfig = require('../../jest.config.base');
const package = require('./package');

module.exports = {
    ...baseConfig,
    displayName: package.name,
    setupFilesAfterEnv: ['jest-extended', `<rootDir>/src/jest-setup.ts`],
};
