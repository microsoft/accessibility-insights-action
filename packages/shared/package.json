{
    "name": "@accessibility-insights-action/shared",
    "private": true,
    "version": "1.0.0",
    "description": "This project welcomes contributions and suggestions.  Most contributions require you to agree to a\r Contributor License Agreement (CLA) declaring that you have the right to, and actually do, grant us\r the rights to use your contribution. For details, visit https://cla.opensource.microsoft.com.",
    "main": "dist/index.js",
    "scripts": {
        "build": "tsc & echo",
        "cbuild": "npm-run-all --serial clean build",
        "clean": "rimraf dist test-results accessibility-insights-action/dist",
        "lint:check": "eslint -c ../../.eslintrc.js \"src/**/*.{js,ts}\"",
        "lint:fix": "eslint -c ../../.eslintrc.js \"src/**/*.{js,ts}\" --quiet --fix",
        "test": "jest",
        "watch:test": "jest --watch"
    },
    "repository": {
        "type": "git",
        "url": "git+https://github.com/microsoft/accessibility-insights-action.git"
    },
    "keywords": [
        "actions"
    ],
    "author": "Microsoft",
    "license": "MIT",
    "bugs": {
        "url": "https://github.com/microsoft/accessibility-insights-action/issues"
    },
    "homepage": "https://github.com/microsoft/accessibility-insights-action#readme",
    "dependencies": {
        "@types/react": "^16.14",
        "@types/react-dom": "^16.9",
        "accessibility-insights-report": "4.7.3",
        "accessibility-insights-scan": "2.5.1",
        "axe-core": "4.7.2",
        "express": "^4.19.2",
        "filenamify-url": "^3.1.0",
        "get-port": "^7.1.0",
        "inversify": "6.0.1",
        "lodash": "^4.17.21",
        "marked": "^9.1.4",
        "marked-terminal": "^7.0.0",
        "normalize-path": "^3.0.0",
        "pony-cause": "^2.1.10",
        "reflect-metadata": "^0.2.1",
        "serialize-error": "^11.0.3",
        "serve-static": "^1.15.0"
    },
    "devDependencies": {
        "@types/express": "^4.17.21",
        "@types/jest": "^29.5.12",
        "@types/lodash": "^4.17.0",
        "@types/marked": "^5.0.2",
        "@types/marked-terminal": "^6.0.1",
        "@types/normalize-path": "^3.0.2",
        "@types/puppeteer-core": "^7.0.4",
        "@types/serve-static": "^1.15.5",
        "@typescript-eslint/eslint-plugin": "^6.18.1",
        "@typescript-eslint/parser": "^6.18.1",
        "eslint": "^8.57.0",
        "eslint-plugin-security": "^1.7.1",
        "fork-ts-checker-webpack-plugin": "^9.0.2",
        "jest": "^29.7.0",
        "jest-extended": "^4.0.2",
        "jest-file-snapshot": "^0.5.0",
        "jest-junit": "^16.0.0",
        "mockdate": "^3.0.5",
        "npm-run-all": "^4.1.5",
        "prettier": "^3.2.5",
        "rimraf": "^5.0.5",
        "ts-jest": "^29.1.2",
        "ts-loader": "^9.5.1",
        "typemoq": "^2.1.0",
        "typescript": "^5.4.2"
    }
}
