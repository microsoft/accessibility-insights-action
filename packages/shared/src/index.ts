// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

export { setupSharedIocContainer } from './ioc/setup-ioc-container';
export { Logger } from './logger/logger';
export { Scanner } from './scanner/scanner';
export { iocTypes } from './ioc/ioc-types';
export { ProgressReporter } from './progress-reporter/progress-reporter';
export { ReportMarkdownConvertor } from './mark-down/report-markdown-convertor';
export { productTitle } from './mark-down/markdown-formatter';
export { disclaimerText } from './content/mark-down-strings';
export { checkRunDetailsTitle, checkRunName } from './content/strings';
export { TaskConfig } from './task-config';
export { BaselineInfo } from './baseline-info';
export { ArtifactsInfoProvider } from './artifacts-info-provider';
export { hookStdout } from './output-hooks/hook-stdout';
export { hookStderr } from './output-hooks/hook-stderr';
