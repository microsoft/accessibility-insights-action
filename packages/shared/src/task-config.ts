// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import { inject, injectable } from 'inversify';
import { iocTypes } from './ioc/ioc-types';

export type TaskInputKey = 'HostingMode' | 'StaticSiteDir' | 'StaticSiteUrlRelativePath' | 'Url' | 'StaticSitePort';
@injectable()
export abstract class TaskConfig {
    constructor(@inject(iocTypes.Process) protected readonly processObj: typeof process) {}
    abstract getReportOutDir(): string;
    abstract getStaticSiteDir(): string | undefined;
    abstract getStaticSiteUrlRelativePath(): string | undefined;
    abstract getSingleWorker(): boolean;
    abstract getBaselineFile(): string | undefined;
    abstract getChromePath(): string | undefined;
    abstract getUrl(): string | undefined;
    abstract getMaxUrls(): number;
    abstract getDiscoveryPatterns(): string | undefined;
    abstract getInputFile(): string | undefined;
    abstract getInputUrls(): string | undefined;
    abstract getScanTimeout(): number;
    abstract getStaticSitePort(): number | undefined;
    abstract getRunId(): number | undefined;
    abstract getHostingMode(): string | undefined;
    abstract getInputName(key: TaskInputKey): string;
    abstract getUsageDocsUrl(): string;
    abstract getServiceAccountName(): string | undefined;
    abstract getServiceAccountPassword(): string | undefined;
    abstract getAuthType(): string | undefined;
}
