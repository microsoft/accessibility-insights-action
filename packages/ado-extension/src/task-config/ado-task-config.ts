// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import * as adoTask from 'azure-pipelines-task-lib/task';
import { inject, injectable } from 'inversify';
import { isEmpty } from 'lodash';
import * as process from 'process';
import { iocTypes, TaskConfig } from '@accessibility-insights-action/shared';
import normalizePath from 'normalize-path';
import { resolve } from 'path';

@injectable()
export class ADOTaskConfig extends TaskConfig {
    constructor(
        @inject(iocTypes.Process) protected readonly processObj: typeof process,
        private readonly adoTaskObj = adoTask,
        private readonly resolvePath: typeof resolve = resolve,
    ) {
        super(processObj);
    }

    public getReportOutDir(): string {
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        return this.getAbsolutePath(this.adoTaskObj.getInput('outputDir'))!;
    }

    public getSiteDir(): string {
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        return this.adoTaskObj.getInput('siteDir')!;
    }

    public getScanUrlRelativePath(): string {
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        return this.adoTaskObj.getInput('scanUrlRelativePath')!;
    }

    public getToken(): string | undefined {
        return this.adoTaskObj.getInput('repoToken') || undefined;
    }

    public getChromePath(): string | undefined {
        let chromePath = this.adoTaskObj.getInput('chromePath') || undefined;
        chromePath = this.getAbsolutePath(chromePath);

        if (isEmpty(chromePath)) {
            chromePath = this.processObj.env.CHROME_BIN;
        }

        return chromePath;
    }

    public getUrl(): string | undefined {
        const value = this.adoTaskObj.getInput('url');

        return isEmpty(value) ? undefined : value;
    }

    public getMaxUrls(): number {
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        return parseInt(this.adoTaskObj.getInput('maxUrls')!);
    }

    public getDiscoveryPatterns(): string | undefined {
        const value = this.adoTaskObj.getInput('discoveryPatterns');

        return isEmpty(value) ? undefined : value;
    }

    public getInputFile(): string | undefined {
        return this.getAbsolutePath(this.adoTaskObj.getInput('inputFile') ?? undefined);
    }

    public getInputUrls(): string | undefined {
        const value = this.adoTaskObj.getInput('inputUrls');

        return isEmpty(value) ? undefined : value;
    }

    public getScanTimeout(): number {
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        return parseInt(this.adoTaskObj.getInput('scanTimeout')!);
    }

    public getLocalhostPort(): number | undefined {
        const value = this.adoTaskObj.getInput('localhostPort');

        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        return isEmpty(value) ? undefined : parseInt(value!, 10);
    }

    public getRunId(): number | undefined {
        const value = this.processObj.env.BUILD_BUILDID;

        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        return isEmpty(value) ? undefined : parseInt(value!, 10);
    }

    public getRepoServiceConnectionName(): string | undefined {
        return this.adoTaskObj.getInput('repoServiceConnectionName') ?? undefined;
    }

    public getFailOnAccessibilityError(): boolean {
        return this.adoTaskObj.getBoolInput('failOnAccessibilityError');
    }

    public getSingleWorker(): boolean {
        return this.adoTaskObj.getBoolInput('singleWorker');
    }

    public getBaselineFile(): string | undefined {
        const value = this.getAbsolutePath(this.adoTaskObj.getInput('baselineFile'));

        return isEmpty(value) ? undefined : value;
    }

    private getAbsolutePath(path: string | undefined): string | undefined {
        if (isEmpty(path)) {
            return undefined;
        }

        const dirname = this.processObj.env.SYSTEM_DEFAULTWORKINGDIRECTORY ?? __dirname;

        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        return normalizePath(this.resolvePath(dirname, normalizePath(path!)));
    }
}
