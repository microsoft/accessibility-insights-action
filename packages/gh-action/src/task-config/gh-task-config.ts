// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import * as actionCore from '@actions/core';
import { inject, injectable } from 'inversify';
import { isEmpty } from 'lodash';
import * as process from 'process';
import { iocTypes, TaskConfig } from '@accessibility-insights-action/shared';
import normalizePath from 'normalize-path';
import { resolve } from 'path';

@injectable()
export class GHTaskConfig extends TaskConfig {
    constructor(
        @inject(iocTypes.Process) protected readonly processObj: typeof process,
        private readonly actionCoreObj = actionCore,
        private readonly resolvePath: typeof resolve = resolve,
    ) {
        super(processObj);
    }

    public getReportOutDir(): string {
        return this.getAbsolutePath(this.actionCoreObj.getInput('output-dir'));
    }

    public getSiteDir(): string {
        return this.actionCoreObj.getInput('site-dir');
    }

    public getScanUrlRelativePath(): string {
        return this.actionCoreObj.getInput('scan-url-relative-path');
    }

    public getToken(): string {
        return this.actionCoreObj.getInput('repo-token');
    }

    public getChromePath(): string {
        let chromePath;
        chromePath = this.getAbsolutePath(this.actionCoreObj.getInput('chrome-path'));

        if (isEmpty(chromePath)) {
            chromePath = this.processObj.env.CHROME_BIN;
        }

        return chromePath;
    }

    public getUrl(): string {
        return this.actionCoreObj.getInput('url');
    }

    public getMaxUrls(): number {
        return parseInt(this.actionCoreObj.getInput('max-urls'));
    }

    public getDiscoveryPatterns(): string {
        const value = this.actionCoreObj.getInput('discovery-patterns');

        return isEmpty(value) ? undefined : value;
    }

    public getInputFile(): string {
        return this.getAbsolutePath(this.actionCoreObj.getInput('input-file'));
    }

    public getInputUrls(): string {
        const value = this.actionCoreObj.getInput('input-urls');

        return isEmpty(value) ? undefined : value;
    }

    public getScanTimeout(): number {
        return parseInt(this.actionCoreObj.getInput('scan-timeout'));
    }

    public getLocalhostPort(): number {
        const value = this.actionCoreObj.getInput('localhost-port');

        return isEmpty(value) ? undefined : parseInt(value, 10);
    }

    public getRunId(): number {
        return parseInt(this.processObj.env.GITHUB_RUN_ID, 10);
    }

    public getSingleWorker(): boolean {
        return this.actionCoreObj.getBooleanInput('single-worker');
    }

    public getBaselineFile(): string | undefined {
        const value = this.getAbsolutePath(this.actionCoreObj.getInput('baseline-file'));

        return isEmpty(value) ? undefined : value;
    }

    private getAbsolutePath(path: string): string {
        if (isEmpty(path)) {
            return undefined;
        }

        const dirname = this.processObj.env.GITHUB_WORKSPACE ?? __dirname;

        return normalizePath(this.resolvePath(dirname, normalizePath(path)));
    }
}
