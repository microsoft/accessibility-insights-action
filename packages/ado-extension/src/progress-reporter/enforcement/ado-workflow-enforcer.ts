// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import { ADOTaskConfig } from '../../task-config/ado-task-config';
import { inject, injectable } from 'inversify';
import { iocTypes, Logger, ProgressReporter } from '@accessibility-insights-action/shared';
import { CombinedReportParameters } from 'accessibility-insights-report';
import { BaselineEvaluation } from 'accessibility-insights-scan';

@injectable()
export class AdoWorkflowEnforcer extends ProgressReporter {
    private scanSucceeded = true;

    constructor(
        @inject(iocTypes.TaskConfig) private readonly adoTaskConfig: ADOTaskConfig,
        @inject(Logger) private readonly logger: Logger,
    ) {
        super();
    }

    public async start(): Promise<void> {
        // We don't do anything for workflow enforcement
    }

    // eslint-disable-next-line @typescript-eslint/require-await
    public async completeRun(combinedReportResult: CombinedReportParameters, baselineEvaluation?: BaselineEvaluation): Promise<void> {
        const baselineFileInput = this.adoTaskConfig.getBaselineFile();

        if (baselineFileInput != null) {
            await this.failIfBaselineNeedsUpdating(baselineFileInput, baselineEvaluation);
        } else {
            await this.failIfAccessibilityErrorExists(combinedReportResult);
        }
    }

    // eslint-disable-next-line @typescript-eslint/require-await
    public async failRun(): Promise<void> {
        this.scanSucceeded = false;
    }

    public didScanSucceed(): Promise<boolean> {
        return Promise.resolve(this.scanSucceeded);
    }

    private async failIfAccessibilityErrorExists(combinedReportResult: CombinedReportParameters): Promise<void> {
        if (this.adoTaskConfig.getFailOnAccessibilityError() && combinedReportResult.results.urlResults.failedUrls > 0) {
            this.logger.logError(
                'Accessibility error(s) were found. To see all failures and scan details, visit the Extensions tab and download the accessibility report.',
            );
            this.logger.logInfo(
                [
                    'To prevent accessibility errors from failing your build, you can:',
                    '* Use a baseline file to avoid failing for known issues, or',
                    '* Set the failOnAccessibilityError task input to false to avoid failing for all issues',
                ].join('\n'),
            );

            await this.failRun();
        }
    }

    private async failIfBaselineNeedsUpdating(baselineFileInput: string, baselineEvaluation?: BaselineEvaluation): Promise<void> {
        if (baselineEvaluation?.suggestedBaselineUpdate) {
            this.logger.logInfo(
                `##vso[task.logissue type=error;sourcepath=${baselineFileInput}] The baseline file does not match scan results.`,
            );
            await this.failRun();
        }
    }
}
