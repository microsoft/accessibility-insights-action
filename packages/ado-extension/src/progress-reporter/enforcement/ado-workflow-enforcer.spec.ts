// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import 'reflect-metadata';

import { Mock, IMock, MockBehavior } from 'typemoq';
import { ADOTaskConfig } from '../../task-config/ado-task-config';
import { CombinedReportParameters } from 'accessibility-insights-report';

import { BaselineEvaluation, BaselineFileContent } from 'accessibility-insights-scan';
import { AdoWorkflowEnforcer } from './ado-workflow-enforcer';
import { RecordingTestLogger } from '@accessibility-insights-action/shared';

describe(AdoWorkflowEnforcer, () => {
    let adoTaskConfigMock: IMock<ADOTaskConfig>;
    let logger: RecordingTestLogger;
    let testSubject: AdoWorkflowEnforcer;

    const reportWithErrors = {
        results: {
            urlResults: {
                failedUrls: 1,
            },
        },
    } as unknown as CombinedReportParameters;
    const reportWithoutErrors = {
        results: {
            urlResults: {
                failedUrls: 0,
            },
        },
    } as unknown as CombinedReportParameters;
    const emptyBaselineEvaluation = {} as BaselineEvaluation;
    const baselineEvaluationWithSuggestedUpdate = {
        suggestedBaselineUpdate: {} as BaselineFileContent,
    } as BaselineEvaluation;
    const baselineEvaluationWithoutSuggestedUpdate = {
        suggestedBaselineUpdate: null,
    } as BaselineEvaluation;

    beforeEach(() => {
        adoTaskConfigMock = Mock.ofType<ADOTaskConfig>(undefined, MockBehavior.Strict);
        logger = new RecordingTestLogger();
        testSubject = new AdoWorkflowEnforcer(adoTaskConfigMock.object, logger);
    });

    describe('completeRun', () => {
        describe('without baseline', () => {
            beforeEach(() => {
                setupBaselineFileInput(undefined);
            });

            it('fails with pinned error log if accessibility error occurs and failOnAccessibilityError=true', async () => {
                setupFailOnAccessibilityErrorInput(true);

                await testSubject.completeRun(reportWithErrors, emptyBaselineEvaluation);

                expect(logger.recordedLogs()).toMatchSnapshot();
                await expect(testSubject.didScanSucceed()).resolves.toBe(false);
            });

            it.each`
                accessibilityErrorExists | failOnAccessibilityError
                ${true}                  | ${false}
                ${false}                 | ${false}
                ${false}                 | ${true}
            `(
                'succeeds silently when accessibilityErrorExists=$accessibilityErrorExists and failOnAccessibilityError=$failOnAccessibilityError',
                async ({ accessibilityErrorExists, failOnAccessibilityError }) => {
                    setupFailOnAccessibilityErrorInput(failOnAccessibilityError);

                    const report = accessibilityErrorExists ? reportWithErrors : reportWithoutErrors;
                    await testSubject.completeRun(report, emptyBaselineEvaluation);

                    expect(logger.recordedLogs()).toStrictEqual([]);
                    await expect(testSubject.didScanSucceed()).resolves.toBe(true);
                },
            );
        });

        // All baseline behavior should be independent of failOnAccessibilityError
        describe.each([true, false])('with baseline and failOnAccessibilityError=%p', (failOnAccessibilityError) => {
            beforeEach(() => {
                setupBaselineFileInput('/some/file');
                setupFailOnAccessibilityErrorInput(failOnAccessibilityError);
            });

            it('fails with pinned error log if baseline needs to be updated', async () => {
                await testSubject.completeRun(reportWithErrors, baselineEvaluationWithSuggestedUpdate);

                expect(logger.recordedLogs()).toMatchSnapshot();
                await expect(testSubject.didScanSucceed()).resolves.toBe(false);
            });

            it('succeeds silently if baseline does not need update', async () => {
                await testSubject.completeRun(reportWithErrors, baselineEvaluationWithoutSuggestedUpdate);

                expect(logger.recordedLogs()).toStrictEqual([]);
                await expect(testSubject.didScanSucceed()).resolves.toBe(true);
            });
        });
    });

    describe('didScanSucceed', () => {
        it('returns true by default', async () => {
            await expect(testSubject.didScanSucceed()).resolves.toBe(true);
        });

        it('returns false after failRun() is called', async () => {
            await testSubject.failRun();
            await expect(testSubject.didScanSucceed()).resolves.toBe(false);
        });
    });

    const setupFailOnAccessibilityErrorInput = (fail: boolean) => {
        adoTaskConfigMock.setup((o) => o.getFailOnAccessibilityError()).returns(() => fail);
    };

    const setupBaselineFileInput = (baselineFile: undefined | string) => {
        adoTaskConfigMock.setup((o) => o.getBaselineFile()).returns(() => baselineFile);
    };
});
