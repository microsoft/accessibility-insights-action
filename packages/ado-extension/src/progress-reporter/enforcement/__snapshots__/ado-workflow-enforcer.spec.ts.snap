// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdoWorkflowEnforcer completeRun with baseline and failOnAccessibilityError=false fails with pinned error log if baseline needs to be updated 1`] = `
Array [
  "[info] ##vso[task.logissue type=error;sourcepath=/some/file] The baseline file does not match scan results.",
]
`;

exports[`AdoWorkflowEnforcer completeRun with baseline and failOnAccessibilityError=true fails with pinned error log if baseline needs to be updated 1`] = `
Array [
  "[info] ##vso[task.logissue type=error;sourcepath=/some/file] The baseline file does not match scan results.",
]
`;

exports[`AdoWorkflowEnforcer completeRun without baseline fails with pinned error log if accessibility error occurs and failOnAccessibilityError=true 1`] = `
Array [
  "[error] Accessibility error(s) were found. To see all failures and scan details, visit the Extensions tab and download the accessibility report.",
  "[info] To prevent accessibility errors from failing your build, you can:
* Use a baseline file to avoid failing for known issues, or
* Set the failOnAccessibilityError task input to false to avoid failing for all issues",
]
`;
